app/jobs/submission_management_charge_calculation_job.rb

```diff
@@ -3,13 +3,52 @@
 # marks it as ready for review
 #
 class SubmissionManagementChargeCalculationJob < ApplicationJob
-  def perform(submission)
-    framework_definition = submission.framework.definition
+  class DefinitionSourceChanged < StandardError; end
 
-    submission.entries.invoices.find_each do |entry|
+  discard_on(DefinitionSourceChanged) do |job, _exception|
+    handle_definition_source_changed(job)
+  end
+
+  def perform(submission, definition_source_at_enqueue_time)
```

Comment:
I might be worrying about nothing here, but I tend to avoid putting large blobs of data that aren't identifiers of some kind onto the job queue. I understand the need to use the same FDL definition when the job is run as existed at the time it was enqueued though, and I was wondering if one of these options would be cleaner:

- Put the FDL definition on the job queue, and actually use that definition to execute the job, rather than loading the framework from the database
- Put a hash of the FDL definition on the queue and use that to validate the definition loaded from the database

I think I prefer the "reject the job if the definition changed" approach as that really means that the return the supplier has provided is not actually considered valid any more.

Reply:
Yeah, I had a similar feeling that it might not be a great idea to put the whole definition in the job metadata. The only reason I decided to do it instead of a hash was so that I could spit out as informative an error message as possible, showing the before/after definition. But if we can live without this then happy to change it to a hash. What do you think?

As for 
> Put the FDL definition on the job queue, and actually use that definition to execute the job, 
rather than loading the framework from the database

I [thought about it and decided it against it](https://trello.com/c/ym3SHha1/1140-allow-editing-of-fdl-for-published-frameworks-in-the-admin-app#comment-5d55863a9f53947f0a4956f6) since it would essentiallly be an ad-hoc framework versioning system, which is something we decided we didn't want.

Reply:
> The only reason I decided to do it instead of a hash was so that I could spit out as informative an error message as possible

That's a good point that I hadn't considered. Is the error message for developers/maintainers or end users? If we go down this route a possible future improvement is to show the diff of the two strings in the error message.

Reply:
It's for developers - it appears in the Rollbar error message.